wrappers/services.go

```diff
@@ -75,30 +76,107 @@ func generateSnapServiceFile(app *snap.AppInfo, opts *AddSnapServicesOptions) ([
 	return genServiceFile(app, opts)
 }
 
+// max returns the maximum of two integers. Why is this
+// not provided by golang?
+func max(a, b int) int {
+	if a < b {
+		return b
+	}
+	return a
+}
+
+func min(a, b int) int {
+	if b < a {
+		return b
+	}
+	return a
+}
+
+func formatCpuGroupSlice(grp *quota.Group) string {
+	if grp.CpuLimit == nil {
+		return ""
+	}
+
+	// calculateSystemdCPULimit calculates the number of cpus and the allowed percentage
+	// to the systemd specific format. The CPUQuota setting is only available since systemd 213
+	calculateSystemdCPULimit := func() string {
+		if grp.CpuLimit.Count == 0 && grp.CpuLimit.Percentage == 0 {
+			return ""
+		}
+
+		cpuQuotaSnap := max(grp.CpuLimit.Count, 1) * grp.CpuLimit.Percentage
+		cpuQuotaMax := runtime.NumCPU() * 100
+		return fmt.Sprintf("CPUQuota=%d%%\n", min(cpuQuotaSnap, cpuQuotaMax))
+	}
+
+	// getAllowedCpusValue converts allowed CPUs array to a comma-delimited
+	// string that systemd expects it to be in. If the array is empty then
+	// an empty string is returned
+	getAllowedCpusValue := func() string {
+		if len(grp.CpuLimit.AllowedCPUs) == 0 {
+			return ""
+		}
+
+		allowedCpusValue := strings.Trim(strings.Join(strings.Fields(fmt.Sprint(grp.CpuLimit.AllowedCPUs)), ","), "[]")
+		return fmt.Sprintf("AllowedCPUs=%s\n", allowedCpusValue)
+	}
+
+	// AllowedCpus is only available for cgroupsv2
+	template := `# Always enable cpu accounting, so the following cpu quota options have an effect
+CPUAccounting=true
+`
+	return fmt.Sprint(template, calculateSystemdCPULimit(), getAllowedCpusValue(), "\n")
+}
+
+func formatMemoryGroupSlice(grp *quota.Group) string {
+	if grp.MemoryLimit == 0 {
+		return ""
+	}
+
+	template := `# Always enable memory accounting otherwise the MemoryMax setting does nothing.
+MemoryAccounting=true
+MemoryMax=%[1]d
+# for compatibility with older versions of systemd
+MemoryLimit=%[1]d
+
+`
+	return fmt.Sprintf(template, grp.MemoryLimit)
+}
+
+func formatTaskGroupSlice(grp *quota.Group) string {
+	if grp.TaskLimit == 0 {
+		return `# Always enable task accounting in order to be able to count the processes/
+# threads, etc for a slice
+TasksAccounting=true
+`
+	}
+
+	template := `# Always enable task accounting in order to be able to count the processes/
+# threads, etc for a slice
+TasksAccounting=true
+TasksMax=%d
+`
+	return fmt.Sprintf(template, grp.TaskLimit)
+}
```

Comment:
I would prefer to only have the template defined once
```suggestion
func formatTaskGroupSlice(grp *quota.Group) string {
	template := `# Always enable task accounting in order to be able to count the processes/
# threads, etc for a slice
TasksAccounting=true
%s
`
	taskLimitStr := "" 
	if grp.TaskLimit != 0 {
		taskLimitStr := fmt.Sprintf("TasksMax=%d",grp.TaskLimit)
	}

	return fmt.Sprintf(template, taskLimitStr)
}
```

Reply:
Yes, definately a good point, I'll get it done